### ISSUE TITLE
Wrong start & end times of warm up period experienced during live trading with Bitfinex

### ISSUE THREAD

**Issue by lucasbrynte** [QuantConnect/Lean#6592]
#### Expected Behavior
1. The extent of the warm up period should equal the configured bar count, e.g. `self.SetWarmupUp(6*24, Resolution.Hour)` should result in a 6 day warm up period.
2. The `self.IsWarmingUp` flag should switch from False to True at the exact time of transition from historical to live data.

#### Actual Behavior
1. The extent of the warm up period is capped at 5 days. The behavior is the same for `Resolution.Minute` and `Resolution.Daily` - the period never exceeds 5 days of historical data. The "truncation" of data happens silently, i.e. without any warning or reason provided. While the warm up period is truncated, a direct history request does not seem to be affected by the issue (the line `self.History(['TESTBTCTESTUSD'], 6*24, Resolution.Hour)` in the provided example successfully retrieves 6 days historical data).
2. The `self.IsWarmingUp` flag switches from False to True too early - a couple of bars before actual live data starts coming in. That is, the last couple of historical trade bars arrive as they should, but they do so after the `self.IsWarmingUp` flag has switched from False to True. More precisely, the algorithm is executed at 220829, 14.54 UTC, and the first bar of live data is received at 15.00 UTC. Nevertheless, as can be seen from the OnData() printouts, starting from 11.00 UTC (the same day), `self.IsWarmingUp` is False. I imagine the behavior can be especially problematic if there are sections of code, conditioned with `if not self.IsWarmingUp`, in which orders may be placed. Then, orders may potentially be placed based on historical data, despite not intended.

#### Reproducing the Problem
The following example algorithm and corresponding logged output illustrates the issue.
[main.py.txt](https://github.com/QuantConnect/Lean/files/9446052/main.py.txt)

### Log
[log.txt](https://github.com/QuantConnect/Lean/files/9446066/log.txt)
Direct request of historical data works as intended:
![image](https://user-images.githubusercontent.com/9484028/187244789-ebf3847d-05ee-4be6-8c83-c4305900362d.png)
Initial OnData() printouts, 5 days back in time:
![image](https://user-images.githubusercontent.com/9484028/187244941-3b60d7d1-5def-4c63-8fae-bf29d7bb431f.png)
Final OnData() printouts, including the historical times 11.00, 12.00, 13.00, and 14.00, at which `self.IsWarmingUp` is False:
![image](https://user-images.githubusercontent.com/9484028/187245632-b6c07574-aacb-466c-81a6-6efc4cbf2980.png)

#### System Information
OS: Manjaro Linux. Linux kernel: 5.4.210-1-MANJARO
Docker image: `quantconnect/lean:14519`.
Lean CLI version: `1.0.103`

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

**Comment by lucasbrynte** [comment]
Btw, although this does not seem to be a practical issue, I would like to check whether I should be concerned by the following error message?
![image](https://user-images.githubusercontent.com/9484028/187246643-aba56101-2651-4c50-82cd-0d6de89f7d84.png)

**Comment by Martin-Molinero** [comment]
Hey @lucasbrynte!

1. Think the setting you are looking for is `Config.GetInt("maximum-warmup-history-days-look-back", 5);`. For warmup, we limit the amount of historical data fetched from the history provider and expect the data to be on disk for older data, fetched like a normal backtest. But increasing that configuration value will allow you to fetch more data from the history provider itself.

2. Warmup ends as soon as we get the first slice with current time, see https://github.com/QuantConnect/Lean/blob/master/Algorithm/QCAlgorithm.cs#L1078 **even** if no realtime data has come through yet, since this could depend on multiple factors, imagine daily resolution, or deploying on a weekend it wouldn't finishing warmup until a few days later. It's expected for warmup to have brought the algorithms data to the latest available, just before the algo started.

3. https://github.com/QuantConnect/Lean/issues/6592#issuecomment-1230532477 This is expected 👌  some history providers do not provide certain types of historical data, like quote information or tick resolution are usually the ones not available.

Let me know if you have any more thoughts, hope it makes sense, will close the issue for now 👍

**Comment by lucasbrynte** [comment]
Hi @Martin-Molinero, and thank you for your response!

I think 1. is mostly resolved, but not 2. entirely - please find my comments on both matters below.

1. Wonderful! By simply increasing this configuration value, it seems I can access a lot more historical data! I have a few comments, mostly to be regarded as feedback on my experience:
  - I am not sure the `"maximum-warmup-history-days-look-back"` parameter is part of the default `lean.json` file. Are perhaps some parameters, including this one, left out intentionally to avoid the config file getting too bloated..?
  - I must say that it was unfortunately silent what was going on. Do you think it would be possible to trigger some kind of warning or error when the requested data is not available, e.g. if my requested warmup period exceeds the `"maximum-warmup-history-days-look-back"` limit while also no data is available on disk?
  - Actually, I have missed a similar feature during backtests. If I run a backtest for a historical period exceeding how much data is available on disk, I would much appreciate a warning or error. I guess a complicating factor would be that sometimes data is missing for valid reasons, e.g. holidays, and I'm unsure whether you have 100% reliable metadata regarding all such valid "off days" or similar. But I think at least a warning would be justified if either the period of missing data is unexpectedly long, or if one has requested a warmup period with a bar count greater than how much data is actually available preceding the `StartDate`.
2. I'm not sure I grasped fully what you meant here, but thanks for pointing me to which part of the code triggers the end of the warm up phase. Also forgive me if I'm wrong, but I think there might be a potential time zone inconsistency causing issue 2. described above. In my live deploy experiments, I don't seem to get this issue if I avoid the `self.SetTimeZone("UTC")` call in the `Initialize()` method. Consider the following lines of code, which I believe are executed in the following order:
  a. First `algorithm.SetLiveMode(true)` is called, causing `algorithm._start` to be initialized, using the default local time zone "New York": https://github.com/QuantConnect/Lean/blob/98fa4464ce93bb00d142e870a53285b0a1bb1399/Engine/Setup/BrokerageSetupHandler.cs#L244
  b. Next `algorithm.Initialize()` is called, potentially leading to a change of time zone (if the user has chosen another time zone). The time zone in which `algorithm._start` is expressed does however remain unchanged, which I believe causes an inconsistency: https://github.com/QuantConnect/Lean/blob/98fa4464ce93bb00d142e870a53285b0a1bb1399/Engine/Setup/BrokerageSetupHandler.cs#L281
  c. Later, `SetFinishedWarmingUp()` is called whenever the following `if` branch is executed, but the condition involves comparing `_start` (NY time zone) with `Time` (user-configured time zone), leading to a potential bias if the time zones are different: https://github.com/QuantConnect/Lean/blob/master/Algorithm/QCAlgorithm.cs#L1076

**Comment by Martin-Molinero** [comment]
Hey @lucasbrynte!
Thanks for looking into it ✨ you were right, we've shipped https://github.com/QuantConnect/Lean/pull/6612 addressing the time zone issue and adding the configuration value into the main config.json so they are easier to find for new users. Hope this addresses the issues.

> I have missed a similar feature during backtests. If I run a backtest for a historical period exceeding how much data is available on disk

We've discussed something similar, adding some type of feedback on missing data and allowing the user to download it with 1 click/enter/option, it's in the pipeline 👍 

Thanks again 🙌
